cmd/logging/logging.go

```diff
@@ -0,0 +1,68 @@
+package logging
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+)
+
+// Config is passed to the binary logging function
+type Config struct {
+	ID        string
+	Namespace string
+	Stdout    io.Reader
+	Stderr    io.Reader
+}
+
+// LoggerFunc is a binary logging function signature
+type LoggerFunc func(context.Context, *Config, func() error) error
+
+// Run runs LoggerFunc
+func Run(fn LoggerFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var errCh = make(chan error, 0)
+
+	sout, _ := winio.DialPipeContext(ctx, os.Getenv("CONTAINER_STDOUT"))
+	serr, _ := winio.DialPipeContext(ctx, os.Getenv("CONTAINER_STDERR"))
+	wait, _ := winio.DialPipeContext(ctx, os.Getenv("CONTAINER_WAIT"))
+
+	config := &Config{
+		ID:        os.Getenv("CONTAINER_ID"),
+		Namespace: os.Getenv("CONTAINER_NAMESPACE"),
+		Stdout:    sout,
+		Stderr:    serr,
+	}
+
+	// Write to wait pipe
+	ready := func() error {
+		wait.Write([]byte("#"))
+		return wait.Close()
+	}
+
+	go func() {
+		if err := fn(ctx, config, ready); err != nil {
+			errCh <- err
+			return
+		}
+		errCh <- nil
+	}()
+
+	for {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				os.Exit(1)
+			}
+			os.Exit(0)
+		default:
+			time.Sleep(500 * time.Millisecond)
+		}
+	}
+}
```

Comment:
The above goroutine that calls the LoggerFunc seemingly is the exit case for this entire function and it looks like it's only run once. Can we drop the select and for loop and just wait for the first value sent down errCh?

Reply:
yeah, makes sense